framework/src/service/ServiceRegistrationBase.cpp

```diff
@@ -39,61 +40,50 @@ US_MSVC_DISABLE_WARNING(4503) // decorated name length exceeded, name was trunca
 namespace cppmicroservices
 {
 
-    ServiceRegistrationBase::ServiceRegistrationBase() {}
+    ServiceRegistrationBase::ServiceRegistrationBase() = default;
 
-    ServiceRegistrationBase::ServiceRegistrationBase(ServiceRegistrationBase const& reg) : d(reg.d)
-    {
-        if (d)
-            ++d->ref;
-    }
+    ServiceRegistrationBase::ServiceRegistrationBase(ServiceRegistrationBase const& reg) = default;
 
-    ServiceRegistrationBase::ServiceRegistrationBase(ServiceRegistrationBase&& reg) noexcept : d(nullptr)
-    {
-        std::swap(d, reg.d);
-    }
+    ServiceRegistrationBase::ServiceRegistrationBase(ServiceRegistrationBase&& reg) noexcept = default;
 
-    ServiceRegistrationBase::ServiceRegistrationBase(ServiceRegistrationBasePrivate* registrationPrivate)
+    ServiceRegistrationBase::ServiceRegistrationBase(
+        std::shared_ptr<ServiceRegistrationBasePrivate> registrationPrivate)
         : d(registrationPrivate)
     {
-        if (d)
-            ++d->ref;
     }
 
     ServiceRegistrationBase::ServiceRegistrationBase(BundlePrivate* bundle,
                                                      InterfaceMapConstPtr const& service,
                                                      Properties&& props)
-        : d(new ServiceRegistrationBasePrivate(bundle, service, std::move(props)))
+        : d(std::make_shared<ServiceRegistrationBasePrivate>(bundle, service, std::move(props)))
     {
+        d->CreateReference();
```

Comment:
What does this addition do?

Reply:
To create the ServiceReference, you need to pass in a ServiceRegistrationBasePrivate shared_ptr, but that cannot be passed until the ServiceRegistrationBasePrivate object has been fully instantiated. So first we make the shared_ptr<ServiceRegistrationBasePrivate> and then we call createReference once that shared_ptr is finished being made to call shared_from_this when making the ServiceReferenceBase object for that ServiceRegistrationBasePrivate.

Reply:
Please add that info to a comment  there.